File: backend/app/schemas/adoption.py
from datetime import datetime

from pydantic import BaseModel, model_validator

from .enums import ApplicationStatus


class AdoptionApplicationCreate(BaseModel):
    pet_id: int
    message: str | None = None


class AdoptionApplicationStatusUpdate(BaseModel):
    status: ApplicationStatus


class AdoptionApplicationOut(BaseModel):
    id: int
    user_id: int
    pet_id: int
    message: str | None = None
    status: ApplicationStatus
    created_at: datetime

    model_config = {"from_attributes": True}


class AdminApplicationOut(AdoptionApplicationOut):
    applicant_name: str = ""
    applicant_email: str = ""
    pet_name: str = ""
    pet_photo_url: str | None = None

    @model_validator(mode="before")
    @classmethod
    def flatten_relations(cls, data):
        if hasattr(data, "user") and data.user is not None:
            data.applicant_name = data.user.full_name
            data.applicant_email = data.user.email
        if hasattr(data, "pet") and data.pet is not None:
            data.pet_name = data.pet.name
            data.pet_photo_url = data.pet.photo_url
        return data

File: frontend/lib/features/adoptions/models/adoption.dart
class Adoption {
  const Adoption({
    required this.id,
    required this.petId,
    required this.petName,
    required this.petImageUrl,
    required this.status,
    required this.createdAt,
  });

  final int id;
  final int petId;
  final String petName;
  final String? petImageUrl;
  final String status;
  final DateTime createdAt;

  factory Adoption.fromJson(Map<String, dynamic> json) {
    return Adoption(
      id: json['id'] as int,
      petId: json['pet_id'] as int,
      petName: json['pet_name'] as String? ?? '',
      petImageUrl: json['pet_photo_url'] as String?,
      status: json['status'] as String? ?? 'pending',
      createdAt: DateTime.parse(json['created_at'] as String),
    );
  }
}
